src/circle_sel.c

```diff
@@ -0,0 +1,218 @@
+		return DEFAULT_SCIRCLE_SEL;
+	}
+	if (!IsA(lsecond(ofunc->args), Const))
+	{
+		ereport(DEBUG1, (errmsg("<@ selectivity called on scircle() with non-const 2nd arg")));
+		return DEFAULT_SCIRCLE_SEL;
+	}
+	arg2 = (Const *) lsecond(ofunc->args);
+	if (arg2->consttype != FLOAT8OID)
+	{
+		ereport(DEBUG1, (errmsg("<@ selectivity called on scircle() with non-float8 2nd arg")));
+		return DEFAULT_SCIRCLE_SEL;
+	}
+
+	radius = DatumGetFloat8(arg2->constvalue);
+	selec = spherecircle_area_float(radius) / SPHERE_SURFACE;
+	CLAMP_PROBABILITY(selec);
+
+	return selec;
+}
+
+static double
+spherepoint_in_circle_sel_internal(PG_FUNCTION_ARGS, bool commute, bool negate)
+{
+	PlannerInfo *root = (PlannerInfo *) PG_GETARG_POINTER(0);
+	Oid         operator = PG_GETARG_OID(1);
+	List       *args = (List *) PG_GETARG_POINTER(2);
+	int         varRelid = PG_GETARG_INT32(3);
+	//Oid         collation = PG_GET_COLLATION();
+	VariableStatData vardata;
+	Node       *other;
+	bool        varonleft;
+	double      selec;
+
+	/*
+	 * When asked about <>, we do the estimation using the corresponding =
+	 * operator, then convert to <> via "1.0 - eq_selectivity - nullfrac".
+	 */
+	if (negate)
+	{
+		operator = get_negator(operator);
+		if (!OidIsValid(operator))
+		{
+			/* Use default selectivity (should we raise an error instead?) */
+			return 1.0 - DEFAULT_SCIRCLE_SEL;
+		}
+	}
+
+	/*
+	 * If expression is not variable = something or something = variable, then
+	 * punt and return a default estimate.
+	 */
+	if (!get_restriction_variable(root, args, varRelid,
+								  &vardata, &other, &varonleft))
+		return negate ? (1.0 - DEFAULT_SCIRCLE_SEL) : DEFAULT_SCIRCLE_SEL;
+
+	/*
+	 * We can do a lot better if the something is a constant.  (Note: the
+	 * Const might result from estimation rather than being a simple constant
+	 * in the query.)
+	 * Look only at var op circle_const, not var op point_const.
+	 */
+	if (IsA(other, Const) && varonleft != commute)
+	{
+		Const	*constnode = (Const *) other;
+		SCIRCLE	*c;
+
+		Assert(!constnode->constisnull); /* operators are strict, shouldn't have NULLs here */
+		c = (SCIRCLE *) constnode->constvalue;
+		selec = spherecircle_area_float(c->radius) / SPHERE_SURFACE;
+		CLAMP_PROBABILITY(selec);
+	}
+	/*
+	 * Check for <@ scircle(..., radius)
+	 */
+	else if (IsA(other, FuncExpr) && varonleft != commute)
+	{
+		selec = spherepoint_in_circle_sel_funcexpr(other);
+		// p *((double * )&((*(Const *)((((FuncExpr*) other)->args->elements)[1].ptr_value)).constvalue))
+	}
+	else
+	{
+		ereport(DEBUG1, (errmsg("<@ selectivity not const, other node tag %i", other->type)));
+		selec = DEFAULT_SCIRCLE_SEL;
+	}
+
+	ReleaseVariableStats(vardata);
+
+	return negate ? (1.0 - selec) : selec;
+}
+
+/*
+ * Common code for spherepoint_in_circle_joinsel()
+ */
+static double
+spherepoint_in_circle_joinsel_internal(PG_FUNCTION_ARGS, bool commute)
+{
+#ifdef NOT_USED
```

Comment:
We can remove it, but it would be useful in case this function is ever fully implemented. Fwiw, the original `#ifdef NOT_USED` is actually coming from the PG function where I copied this from (`eqjoinsel`), but there it was just on one of the variables, not all of them.

Reply:
The function itself is still useful, even if it only returns a static selectivity value. We'd just remove the header with variable definitions.

Reply:
I'm not familiar with PostgreSQL internals enough to have a solid opinion. If this is standard boilerplate code for implementing a selectivity function for joins (and it sounds like it is), I'd be inclined to leave it, especially if there's intent to implement it later. That said, if there's no intent to _ever_ implement it, then I agree with removing the `NOT_USED` section, too.

@df7cb : I see in the comments that you recommend that "joins should use the spoint_dwithin function instead." Could you add that recommendation to the documentation and give an example of a join using the spoint_dwithin function there?

Reply:
> @df7cb : I see in the comments that you recommend that "joins should use the spoint_dwithin function instead." Could you add that recommendation to the documentation and give an example of a join using the spoint_dwithin function there?

@esabol it's already in there:

```
Example 6.6.  Efficiently join two tables of points with some fuzziness permitted

sql> SELECT * FROM stars1 JOIN stars2 WHERE spoint_dwithin(stars1.s, stars2.s, 1e-5);
```

Reply:
@esabol yes this is standard boilerplate for join selectivity.

Reply:
> @esabol it's already in there:

Oh! Good.